library-best-practices/glossary.md

```diff
@@ -0,0 +1,87 @@
+    reflective reference.
+
+- **Behavior conflict**: The class's implementation has changed in a way that
+  can break clients at runtime although all signatures remain compatible. For
+  example, if a method that has been returning mutable lists begins returning
+  immutable lists without updating its signature, dependents that mutate the
+  list will fail, possibly far away from the original call site. By contrast, a
+  change in return type from List to ImmutableList would be a linkage conflict.
+  - Opposite: **Behavior-compatible**.
+
+- **Version disagreement**: This is a class of failure reported by build systems
+  with strict version checking on. The main example is dependency convergence
+  failures when using the Maven enforcer plugin, which checks that all branches
+  of a dependency tree agree on the version of every dependency. A failure of
+  dependency convergence doesn't guarantee a linkage or behavior conflict, and
+  the passing of a dependency convergence check doesn't guarantee there will be
+  no linkage or behavior conflicts. Consequently, version disagreement is a poor
+  proxy for real problems (linkage and behavior conflicts).
+  - Dependency convergence will fail even if the dependency tree only disagrees
+    at the level of a patch version (1.10.0 vs 1.10.1) and there is no linkage
+    or behavioral conflict.
+  - It is possible for dependency convergence to pass when a version of a
+    library is chosen at build time which is incompatible with another library
+    in the tree. This is possible by forcing Maven to pick the "wrong" version
+    using something like dependency management or dependency exclusion, which
+    silences the Maven enforcer plugin.
+  - Opposite: **Version agreement**.
+
+- **Version alignment**: Said of the dependency tree of a Maven module. This
+  means that for any dependency of a module in that module's dependency tree,
+  all major Java build systems will select the same version of that dependency.
+  - Major build systems currently include Maven and Gradle.
+  - Direct (first-order) dependencies will trivially comply with this rule in
+    all major build systems, so the real concern is transitive (second-order and
+    higher) dependencies.
+  - Generally, to achieve linkage compatibility (when assuming breaking changes
+    aren't made within major versions), the upper bound of the versions in the
+    tree should be selected.
+    - This should fix all cases of linkage conflicts caused by purely additive
+      functionality.
+  - To achieve version alignment when multiple versions of a transitive
+    dependency are present, a direct dependency on the transitive dependency
+    will need to be added, in order to support Maven's dependency resolution
+    logic (which selects the version with the fewest hops on the dependency
+    tree).
+  - Linkage compatibility is the desired result from version alignment, but not
```

Comment:
version alignment does not guarantee linkage compatibility because linkage errors occur between different artifacts. I suggest removing this bullet point.

Reply:
This is true, but if we remove this note, then people will wonder what the purpose is. It's like eating healthy: it won't guarantee that you will live longer with a higher quality of life, but that is the desired result of doing it.